crates/devices/virtio-console/src/console.rs

```diff
@@ -0,0 +1,516 @@
+use virtio_queue::DescriptorChain;
+use vm_memory::{Bytes, GuestMemory, GuestMemoryError};
+
+/// Console device errors.
+#[derive(Debug)]
+pub enum Error {
+    /// Failed to write slice to guest memory.
+    WriteToGuestFailed(GuestMemoryError),
+    /// Write to output sink failed.
+    WriteToOutputFailed(GuestMemoryError),
+    /// Transmitq descriptor chain length has overflown
+    ChainLengthOverflow,
+    /// Input buffer maximum capacity exceeded
+    BufferCapacityExceeded,
+    /// Read only descriptor that protocol says to write to.
+    UnexpectedReadOnlyDescriptor,
+    /// Write only descriptor that protocol says to read from.
+    UnexpectedWriteOnlyDescriptor,
+    /// The constructor's parameter for maximum capacity is 0
+    InvalidBufferCapacity,
+    /// Output sink flush has not written all bytes.
+    OutputSinkFlushFailed(std::io::Error),
+}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use self::Error::*;
+
+        match self {
+            WriteToGuestFailed(ref err) => {
+                write!(f, "Console failed to write slice to guest memory: {}", err)
+            }
+            WriteToOutputFailed(ref err) => {
+                write!(f, "Console failed to write data to output sink: {}", err)
+            }
+            ChainLengthOverflow => {
+                write!(f, "Console transmitq chain length has overflown usize.")
+            }
+            BufferCapacityExceeded => {
+                write!(
+                    f,
+                    "Console input buffer maximum capacity has been exceeded."
+                )
+            }
+            UnexpectedReadOnlyDescriptor => {
+                write!(f, "Unexpected read-only descriptor.")
+            }
+            UnexpectedWriteOnlyDescriptor => {
+                write!(f, "Unexpected write-only descriptor.")
+            }
+            InvalidBufferCapacity => {
+                write!(f, "The maximum capacity of the input buffer cannot be 0.")
+            }
+            OutputSinkFlushFailed(ref err) => {
+                write!(f, "Output sink flush has not written all bytes: {}", err)
+            }
+        }
+    }
+}
+
+/// Console struct that implements the abstraction of virtio console descriptor chain handling.
+#[derive(Clone, Debug)]
+pub struct Console<T: Write> {
+    /// Buffer that stores data to be sent to the driver.
+    input_buffer: Arc<Mutex<Vec<u8>>>,
+    /// Capacity of the input buffer.
+    capacity: usize,
+    /// Output sink.
+    output: T,
+}
+
+// Default maximum capacity
+// TODO: Find the optimal CAPACITY and MAX_CAPACITY through testing
+const MAX_CAPACITY: usize = 16384;
+const DEFAULT_CAPACITY: usize = 4096;
+
+impl Default for Console<Stdout> {
+    fn default() -> Self {
+        Self::new_with_capacity(DEFAULT_CAPACITY, stdout()).unwrap()
+    }
+}
+
+/// Used to check if two consoles are of the same size.
+impl<T> PartialEq for Console<T>
```

Comment:
We shouldn't implement `PartialEq` and export it in the public interface like this because it doesn't make sense now that we also have the `output` as part of the `Console`. We could have a dummy `PartialEq` implemented for testing purposes, but this should not be part of the public interface. Also, for that one, we should check that the input is also the same. I'm sorry if my initial comment about the Rust API guidelines confused you.